[env] Add CREATE_OR_OPEN and rename existing open modes

lingbin opened the pull request:
The upcoming patch will use CREATE_OR_OPEN mode

This patch also remove virtual dtors to cpp file.

lingbin committed:
[env] Add CREATE_OR_OPEN and rename existing open modes

The upcoming patch will use CREATE_OR_OPEN mode

This patch also remove virtual dtors to cpp file.

lingbin committed:
Fix UT

imay commented on be/src/env/env.h:
```diff
@@ -29,19 +29,21 @@ class Env {
+        CREATE_OR_OPEN_WITH_TRUNCATE,
+        CREATE_OR_OPEN,
+        MUST_CREATE,
+        MUST_EXIST
```
`CREATE_OR_OPEN` starts with a verb, and MUST_CREATE is not. we should keep it consistent.

lingbin committed:
Move the dtors back to env.h

Generally, placing the dtor in an `.h` file(inline) or in a `cpp` file
depends on the trade-off between code expansion and function call overhead.
The code expansion rate is closely related to the number of class members
and the inheritance level.

For the several classes here: `Env`, `ReadableFile`, and `WritableFile`
have no members and are the top level of the inheritance hierarchy, But
for now I have no obvious evidence to prove that make their dtors inline
will cause serious code expansion and more instruction cache-misses,
even if there are thousands of `ReadableFile` objects kept being created
and released during running.

imay reviewed (approved):
LGTM
